File: sdk/flutter/test/copy_test.dart
// Unit tests for locale resolution, MothCopy resolution/interpolation and the
// bundled fallback catalog — no server needed.
import 'dart:ui';

import 'package:flutter_test/flutter_test.dart';
import 'package:moth_auth/moth_auth.dart';

void main() {
  group('locale', () {
    test('BCP-47 tag for language and language-region', () {
      expect(mothLanguageTag(const Locale('fr')), 'fr');
      expect(mothLanguageTag(const Locale('fr', 'CA')), 'fr-CA');
    });

    test('parses a tag back to a Locale, tolerating both separators', () {
      expect(mothLocaleFromTag('fr'), const Locale('fr'));
      expect(mothLocaleFromTag('fr-CA'), const Locale('fr', 'CA'));
      expect(mothLocaleFromTag('fr_CA'), const Locale('fr', 'CA'));
      expect(mothLocaleFromTag(''), const Locale('en'));
      expect(
        mothLocaleFromTag('zh-Hant-TW'),
        Locale.fromSubtags(
          languageCode: 'zh',
          scriptCode: 'Hant',
          countryCode: 'TW',
        ),
      );
    });

    test('MothConfig.locale overrides the device locale on the client', () {
      final overridden = MothClient(
        MothConfig(
          endpoint: _dummy,
          publishableKey: 'pk',
          locale: const Locale('fr', 'CA'),
        ),
        tokenStore: InMemoryTokenStore(),
      );
      addTearDown(overridden.dispose);
      expect(mothLanguageTag(overridden.currentLocale), 'fr-CA');

      final device = MothClient(
        MothConfig(endpoint: _dummy, publishableKey: 'pk'),
        tokenStore: InMemoryTokenStore(),
      );
      addTearDown(device.dispose);
      // Falls back to the engine's device locale (non-empty in the test host).
      expect(mothLanguageTag(device.currentLocale), isNotEmpty);
      expect(device.currentLocale, mothDeviceLocale());
    });
  });

  group('MothCopy.value', () {
    test('server override wins, then bundled, then English', () {
      const copy = MothCopy(
        locale: Locale('fr'),
        revisionId: 'r1',
        messages: {'sign_in.title': 'Connexion perso'},
      );
      // Server-delivered override.
      expect(copy.value('sign_in.title'), 'Connexion perso');
      // Key not in the override → bundled French default.
      expect(copy.value('sign_in.submit'), 'Se connecter');
    });

    test('a non-bundled locale falls back to English per key', () {
      const copy = MothCopy(locale: Locale('xx'), revisionId: '', messages: {});
      expect(copy.value('sign_in.title'), 'Sign in');
      expect(copy.value('paywall.cta'), 'Continue');
    });

    test('interpolates {placeholders} from vars, leaving unknown ones', () {
      final copy = MothCopy.bundled(const Locale('fr'));
      expect(
        copy.value('sign_in.subtitle', vars: {'app': 'Acme'}),
        'Bon retour sur Acme.',
      );
      // No vars: the placeholder is left intact.
      expect(copy.value('sign_in.subtitle'), 'Bon retour sur {app}.');
    });

    test('an unknown key returns the key itself', () {
      final copy = MothCopy.bundled(const Locale('en'));
      expect(copy.value('nope.missing'), 'nope.missing');
    });

    test(
      'the bundled floor renders localized copy with no server messages',
      () {
        final copy = MothCopy.bundled(const Locale('de'));
        expect(copy.revisionId, '');
        expect(copy.value('sign_in.title'), 'Anmelden');
        expect(copy.value('paywall.cta'), 'Weiter');
      },
    );
  });

  group('bundled catalog', () {
    test('every SDK-screen key resolves non-empty in every bundled locale', () {
      final keys = bundledCopy(const Locale('en')).keys;
      expect(keys, isNotEmpty);
      for (final code in mothBundledLocales) {
        final map = bundledCopy(Locale(code));
        for (final key in keys) {
          expect(map[key], isNotNull, reason: '$code missing $key');
          expect(map[key], isNotEmpty, reason: '$code empty for $key');
        }
      }
    });

    test('non-English locales carry real translations, not English', () {
      expect(bundledCopy(const Locale('fr'))['sign_in.title'], 'Connexion');
      expect(bundledCopy(const Locale('ja'))['paywall.cta'], '続ける');
    });

    test('an unbundled locale returns the full English map', () {
      final map = bundledCopy(const Locale('xx'));
      expect(map['sign_in.title'], 'Sign in');
      expect(map['sign_up.legal'], startsWith('By continuing'));
    });

    test('covers the SDK screens plus the shared error group', () {
      final screens = bundledCopy(
        const Locale('en'),
      ).keys.map((k) => k.split('.').first).toSet();
      expect(screens, {
        'sign_in',
        'sign_up',
        'password_reset',
        'paywall',
        'error',
      });
    });
  });
}

final Uri _dummy = Uri.parse('http://localhost:1');

File: sdk/react/src/core/copy.test.ts
import { describe, expect, it } from 'vitest'
import { MothCopy, languageOf } from './copy.js'
import { bundledCopy, mothBundledLocales } from './i18n/bundledCopy.js'

describe('MothCopy resolution', () => {
  it('server override wins over the bundled catalog', () => {
    const copy = new MothCopy('en', 'rev', {
      'sign_in.title': 'Enter the lair',
    })
    expect(copy.value('sign_in.title')).toBe('Enter the lair')
  })

  it('falls back to the bundled locale, then English, then the key', () => {
    const fr = new MothCopy('fr', 'rev', {})
    expect(fr.value('sign_in.title')).toBe('Connexion')
    // A bundled locale mirrors the full English key set.
    expect(fr.value('paywall.most_popular')).toBe('Le plus populaire')
    // Unbundled locale: English.
    const xx = new MothCopy('xx')
    expect(xx.value('sign_in.title')).toBe('Sign in')
    // Unknown key: the key itself.
    expect(xx.value('does.not.exist')).toBe('does.not.exist')
  })

  it('region tags resolve through their language', () => {
    const frCa = MothCopy.bundled('fr-CA')
    expect(frCa.value('sign_in.title')).toBe('Connexion')
    expect(languageOf('fr-CA')).toBe('fr')
    expect(languageOf('zh_Hant_TW')).toBe('zh')
  })

  it('substitutes {name} placeholders literally, leaving unknown ones', () => {
    const copy = new MothCopy('en', 'rev', {
      greeting: 'Hello {name}, welcome to {app} {missing}',
    })
    expect(copy.value('greeting', { name: 'Ada', app: 'TestApp' })).toBe(
      'Hello Ada, welcome to TestApp {missing}',
    )
  })

  it('fills {app} in the bundled floor (the server interpolates its own)', () => {
    const copy = MothCopy.bundled('en')
    expect(copy.value('sign_in.subtitle', { app: 'TestApp' })).toBe(
      'Welcome back to TestApp.',
    )
    expect(
      copy.value('sign_up.password_too_short', { count: '8' }),
    ).toBe('Use at least 8 characters')
  })
})

describe('bundled catalog', () => {
  it('bundles the documented locales, English first', () => {
    expect(mothBundledLocales[0]).toBe('en')
    expect([...mothBundledLocales]).toEqual(
      expect.arrayContaining(['en', 'fr', 'de', 'es', 'pt', 'it', 'ja']),
    )
  })

  it('every locale resolves every English key non-empty', () => {
    const en = bundledCopy('en')
    const keys = Object.keys(en)
    expect(keys.length).toBeGreaterThan(70)
    for (const locale of mothBundledLocales) {
      const catalog = bundledCopy(locale)
      for (const key of keys) {
        expect(catalog[key], `${locale}:${key}`).toBeTruthy()
      }
    }
  })

  it('includes the web-only unavailable-on-web key', () => {
    expect(bundledCopy('en')['paywall.unavailable_web']).toContain('web')
    expect(bundledCopy('ja')['paywall.unavailable_web']).toBe(
      'ウェブでは購入できません',
    )
  })
})
